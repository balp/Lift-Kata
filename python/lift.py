from enum import Enum
from typing import List
from attr import dataclass


class Direction(Enum):
    UP = 0
    DOWN = 1


@dataclass
class Call:
    floor: int
    direction: Direction


class Lift:

    def __init__(self, id, floor, doors_open=False, requested_floors=None):
        self.id = id
        self.floor = floor
        self.doors_open = doors_open
        self.requested_floors = list(requested_floors) if requested_floors else []

    def request(self, floor: int) -> None:
        self.requested_floors.append(floor)

    def tick(self):
        if self.floor == self.requested_floors[0]:
            self.doors_open = True
            self.requested_floors.pop(0)
        elif self.doors_open:
            self.doors_open = False
        else:
            self.floor = self.requested_floors[0]

class LiftSystem:
    def __init__(self, floors=None, calls=None, lifts=None):
        self.floors = list(floors) if floors else []
        self.calls = list(calls) if calls else []
        self.lifts: List[Lift] = list(lifts) if lifts else []

    def calls_for(self, floor):
        return [c for c in self.calls if c.floor == floor]

    def tick(self):
        for lift in self.lifts:
            lift.tick()


