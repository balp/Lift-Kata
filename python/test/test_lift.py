from approvaltests import verify

from lift import Lift, LiftSystem
from lift_printers import print_lifts


def test_request_to_move():
    lifts, lifts_before_tick = set_up_simple_lift_system(start_floor=2, request_floor=1, doors_open=False)
    lifts.tick()
    lifts_after_tick = print_lifts(lifts)
    _verify_lifts(lifts_after_tick, lifts_before_tick)


def test_on_requested_floor_open_door():
    lifts, lifts_before_tick = set_up_simple_lift_system(start_floor=1, request_floor=1, doors_open=False)
    lifts.tick()
    lifts_after_tick = print_lifts(lifts)
    _verify_lifts(lifts_after_tick, lifts_before_tick)


def test_dont_move_on_request_with_open_doors():
    lifts, lifts_before_tick = set_up_simple_lift_system(start_floor=1, request_floor=2, doors_open=True)
    lifts.tick()
    lifts_after_tick = print_lifts(lifts)
    _verify_lifts(lifts_after_tick, lifts_before_tick)


def set_up_simple_lift_system(start_floor, request_floor, doors_open):
    lift_a = Lift("A", start_floor)
    lift_a.request(request_floor)
    lift_a.doors_open = doors_open
    lifts = LiftSystem(floors=[1, 2], lifts=[lift_a])
    lifts_before_tick = print_lifts(lifts)
    return lifts, lifts_before_tick


def _verify_lifts(lifts_after_tick, lifts_before_tick):
    verify(f"""Before:
{lifts_before_tick}

After:

{lifts_after_tick}""")

