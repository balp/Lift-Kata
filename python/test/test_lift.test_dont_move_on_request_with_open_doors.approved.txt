Before:
2      *  2
1     ]A[ 1


After:

2      *  2
1     [A] 1
